File: LeetCode/面试经典 150 题/双指针/11. 盛最多水的容器/11.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    inline int max(const int &a, const int &b) { return a > b ? a : b; }
    inline int min(const int &a, const int &b) { return a < b ? a : b; }
    int maxArea(vector<int>& height) {
        vector<int>& d = height;
        int l = 0, r = d.size() - 1;
        int ans = min(d[l], d[r]) * (r - l);
        while (l < r) {
            ans = max(ans, min(d[l], d[r]) * (r - l));
            if (d[l] > d[r]) { r--; }
            else { l++; }
        }
        return ans;
    }
};

int main() {
    vector<int> x = {1,8,6,2,5,4,8,3,7};
    Solution s;
    auto ans = s.maxArea(x);
    cout << ans << ' ';
    return 0;
}

File: LeetCode/LeetCode 75/双指针/11. 盛最多水的容器/11.java
class Solution {
    int[] h;
    public int maxArea(int[] height) {
        h = height;
        int l = 0, r = h.length - 1, ans = getArea(l, r);
        while (l < r) {
            ans = Math.max(ans, getArea(l, r));
            // move lower one won't skip bigger area
            if (h[l] > h[r]) {
                r--;
            } else {
                l++;
            }
        }
        return ans;
    }
    int getArea(int l, int r) {
        return (r - l) * Math.min(h[l], h[r]);
    }
}